src/renderer.cpp

```diff
@@ -1357,6 +1750,248 @@ void Renderer::setD3DProjection(struct matrix* matrix)
+    worldLightRot.y = dirY;
+    worldLightRot.z = dirZ;
+}
+
+struct point3d Renderer::getWorldLightDir()
+{
+    return worldLightRot;
+}
+
+void Renderer::setLightIntensity(float intensity)
+{
+    internalState.lightData[3] = intensity;
+}
+
+float Renderer::getLightIntensity()
+{
+    return internalState.lightData[3];
+}
+
+void Renderer::setLightColor(float r, float g, float b)
+{
+    internalState.lightData[0] = r;
+    internalState.lightData[1] = g;
+    internalState.lightData[2] = b;
+}
+
+point3d Renderer::getLightColor()
+{
+    struct point3d color = { internalState.lightData[0],
+                             internalState.lightData[1],
+                             internalState.lightData[2] };
+    return color;
+}
+
+void Renderer::setAmbientIntensity(float intensity)
+{
+    internalState.ambientLightData[3] = intensity;
+}
+
+float Renderer::getAmbientIntensity()
+{
+    return internalState.ambientLightData[3];
+}
+
+void Renderer::setAmbientLightColor(float r, float g, float b)
+{
+    internalState.ambientLightData[0] = r;
+    internalState.ambientLightData[1] = g;
+    internalState.ambientLightData[2] = b;
+}
+
+point3d Renderer::getAmbientLightColor()
+{
+    struct point3d color = { internalState.ambientLightData[0],
+                             internalState.ambientLightData[1],
+                             internalState.ambientLightData[2] };
+    return color;
+}
+
+void Renderer::setRoughness(float roughness)
+{
+    internalState.materialData[0] = roughness;
+}
+
+float Renderer::getRoughness()
+{
+    return internalState.materialData[0];
+}
+
+void Renderer::setMetalness(float metalness)
+{
+    internalState.materialData[1] = metalness;
+}
+
+float Renderer::getMetalness()
+{
+    return internalState.materialData[1];
+}
+
+void Renderer::setShadowFaceCullingEnabled(bool isEnabled)
+{
+    isShadowMapFaceCullingEnabled = isEnabled;
+}
+
+bool Renderer::isShadowFaceCullingEnabled()
+{
+    return isShadowMapFaceCullingEnabled;
+}
+
+void Renderer::setShadowMapResolution(int resolution)
+{
+    internalState.shadowData[3] = resolution;
+
+    if (bgfx::isValid(shadowMapFrameBuffer))
+    {
+        bgfx::destroy(shadowMapFrameBuffer);
+    }
+
+    shadowMapTexture = bgfx::createTexture2D(
+        resolution,
+        resolution,
+        false,
+        1,
+        bgfx::TextureFormat::D32,
+        BGFX_TEXTURE_RT | BGFX_SAMPLER_COMPARE_LEQUAL | BGFX_SAMPLER_U_CLAMP | BGFX_SAMPLER_V_CLAMP | BGFX_SAMPLER_W_CLAMP
+    );
+
+    shadowMapFrameBuffer = bgfx::createFrameBuffer(
+        1,
+        &shadowMapTexture,
+        true
+    );
+}
+
+int Renderer::getShadowMapResolution()
+{
+    return internalState.shadowData[3];
+}
+
+void Renderer::setShadowConstantBias(float bias)
+{
+    internalState.shadowData[0] = bias;
+}
+
+float Renderer::getShadowConstantBias()
+{
+    return internalState.shadowData[0];
+}
+
+void Renderer::setShadowMapArea(float area)
+{
+    shadowMapArea = area;

```

Comment:
Shouldn't this be part of the `internalState`?

Reply:
I moved this variable to the `internalState`.